analyze/aps-schedule.c

```diff
@@ -0,0 +1,2989 @@
+                        const short ch,
+                        const bool immediate,
+                        bool visit_marker_only,
+                        bool* any) {
+  if (current == NULL)
+    return;
+
+  if (current->child_phase.ph == ph && current->child_phase.ch == ch) {
+    if ((visit_marker_only && IS_VISIT_MARKER(current)) || !visit_marker_only) {
+      *any = true;
+    }
+  } else if (immediate) {
+    return;
+  }
+
+  bool if_true_any = false;
+  bool if_false_any = false;
+
+  if (group_is_local(&current->child_phase) &&
+      if_rule_p(current->cto_instance->fibered_attr.attr)) {
+    followed_by(current->cto_if_true, ph, ch, immediate, visit_marker_only,
+                &if_true_any);
+  }
+
+  followed_by(current->cto_next, ph, ch, immediate, visit_marker_only,
+              &if_false_any);
+
+  *any |= (if_true_any || if_false_any);
+}
+
+/**
+ * Utility function that returns boolean indicating whether there is any <ph,ch>
+ * in augmented dependency graph
+ * @param current CTO_NODE node
+ * @param ph phase
+ * @param ch child number
+ * @return boolean indicating if there is any
+ */
+static bool aug_graph_contains_phase(AUG_GRAPH* aug_graph,
+                                     TOTAL_ORDER_STATE* state,
+                                     CONDITION cond,
+                                     const short ph,
+                                     const short ch) {
+  int i;
+  bool exist_parent_inherited_attr_previous_phase = false;
+  for (i = 0; i < aug_graph->instances.length; i++) {
+    INSTANCE* in = &aug_graph->instances.array[i];
+    CHILD_PHASE* group = &state->instance_groups[in->index];
+
+    if (group->ph == ph && group->ch == ch &&
+        !MERGED_CONDITION_IS_IMPOSSIBLE(instance_condition(in), cond)) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+/**
+ * Utility function used by assert_total_order to check sanity of total order
+ *  1) After visit marker <ph,-1> there should be no attribute
+ *     belonging to <ph,-1> or <-ph,-1>
+ *  2) Immediately before visit marker <ph,ch> should be attribute
+ *     belonging to <-ph,ch>
+ *  3) Or immediately after visit marker <ph,ch> should be attribute belonging
+ *     to <ph,ch>
+ * @param current CTO_NODE node
+ * @param prev_group <ph,ch> group
+ */
+static void total_order_sanity_check(AUG_GRAPH* aug_graph,
+                                     ChunkGraph* chunk_graph,
+                                     CTO_NODE* current,
+                                     CONDITION cond,
+                                     CHILD_PHASE* prev_group,
+                                     CHILD_PHASE* prev_parent,
+                                     TOTAL_ORDER_STATE* state) {
+  if (current == NULL)
+    return;
+
+  CHILD_PHASE* current_group = &current->child_phase;
+
+  if (IS_VISIT_MARKER(current)) {
+    if (current->child_phase.ch == -1) {
+      // End of total order
+      if (current->cto_next == NULL)
+        return;
```

Comment:
I don't understand.  I would think that current->cto_next should be null unless this is the parent INHERITED phase which I think doesn't use a marker.

So why not have an error message if cto_next is *not* null.

Also: after this you have a number of checks.  Those only happen if the cto_next is not null.  Do you not need to check them if cto_next == null?

Reply:
You are correct. Updated the ordering. It prematurely exited the function before making sure before the parent end-of-phase visit marker there were parent synthesized attributes (if any).

This function checks the sanity of CTO nodes. More specifically that code checks if we are currently at end of the parent phase visit marker `<ph,-1>`

- This means there should be a parent synthesized attribute (if any) before the visit marker
- Parent-inherited attributes of the next phase (if any) should be followed by the visit marker. **However**, if this is the last phase then we should check if its followed by parent-inherited attributes of the next phase.


Note that the final CTO node is always the end of the parent visit marker of the last phase of the parent `<max(ph), -1>`